Snakefile

```diff
@@ -0,0 +1,402 @@
+        targets.append(f"data/{database}/trigger-preprocessing.done")
+    if send_notifications:
+        targets.append(f"data/{database}/notify.done")
+    if config.get("fetch_from_database", False):
+        targets.append(f"data/{database}/raw.upload.done")
+    return targets
+
+rule all:
+    input: _compute_targets
+
+#################################################################
+###################### rule definitions #########################
+#################################################################
+
+
+def run_shell_command_n_times(cmd, msg, cleanup_failed_cmd, retry_num=5):
+    attempt = 0
+    while attempt < retry_num:
+        print(f"{msg} attempt number {attempt}")
+        try:
+            subprocess.run(cmd, shell=True, check=True)
+            break
+        except subprocess.CalledProcessError:
+            print("...FAILED")
+            attempt+=1
+            shell("{cleanup_failed_cmd} && sleep 10")
+    else:
+        print(msg + f" has FAILED {retry_num} times. Exiting.")
+        raise Exception("function run_shell_command_n_times has failed")
+
+rule fetch_from_database:
+    message:
+        """Fetching data using the database API"""
+    params:
+        s3_src_bucket = config["s3_src"]
+    output:
+        ndjson = temp("data/{database}.ndjson")
+    run:
+        if wildcards.database=="gisaid":
+            msg = "Fetching from GISAID API"
+            cmd = f"./bin/fetch-from-gisaid > {output.ndjson}"
+        elif wildcards.database=="genbank":
+            msg = "Fetching from GenBank API"
+            cmd = f"./bin/fetch-from-genbank > {output.ndjson}"
+        cleanup_failed_cmd = f"rm {output.ndjson}"
+        run_shell_command_n_times(cmd, msg, cleanup_failed_cmd)
+
+        if send_notifications:
+            shell("./bin/notify-on-record-change {output.ndjson} {params.s3_src_bucket}/gisaid.ndjson.xz {database}")
+
+rule fetch_biosample_from_database:
+    message:
+        """Fetching Biosample from the database"""
+    params:
+        s3_src_bucket = config["s3_src"]
+    output:
+        biosample = "data/biosample.ndjson"
+    run:
+        run_shell_command_n_times(
+            f"./bin/fetch-from-biosample > {output.biosample}",
+            "Fetch BioSample",
+            f"rm {output.biosample}")
+
+def _raw_files_to_upload(wildcards):
+    database = wildcards["database"]
+    files = {f"{database}.ndjson.xz": f"data/{database}.ndjson"}
+    if database=="genbank":
+        files["biosample.ndjson.gz"] = f"data/biosample.ndjson"
+    return files
+
+rule upload_raw_ndjson:
+    input:
+        unpack(_raw_files_to_upload)
+    output:
+        touch("data/{database}/raw.upload.done")
+    params:
+        quiet = "" if send_notifications else "--quiet",
+        s3_bucket = config["s3_dst"]
+    run:
+        for remote, local in input.items():
+            shell("./bin/upload-to-s3 {params.quiet} {local:q} {params.s3_bucket:q}/{remote:q}")
```

Comment:
Post-merge change: This for-loop approach may be historical, but we might consider splitting this out to separate Snakemake rules via wildcards (e.g., `data/{database}/raw.upload.{local}.done`).

Reply:
Agreed! This is a great example of what I hope can be improved now that we use a workflow :) Documented in #239 so resolving here